syntax = "proto2";

package lego.dht.protobuf;

message BootstrapRequest {
	optional string local_ip = 1;
	optional int32 local_port = 2;
	optional int32 nat_type = 3;
	optional bytes node_id = 4;
	optional int32 get_init_msg = 5;
}

message VpnNodeInfo {
    optional bytes ip = 1;
    optional bytes dhkey = 2;
    optional bytes pubkey = 3;
    optional bytes country = 4;
}

message BootstrapResponse {
	optional bytes node_id = 1;
	optional int32 nat_type = 2;
	optional string local_ip = 3;
	optional int32 local_port = 4;
	optional string public_ip = 5;
	optional int32 public_port = 6;
	optional int32 country_code = 7;
	optional bytes version_info = 8;
	repeated VpnNodeInfo route_nodes = 9;
	repeated VpnNodeInfo vpn_nodes = 10;
}

message RefreshNeighborsRequest {
	optional uint32 count = 1;
    optional bytes des_dht_key = 2;
    repeated uint64 bloomfilter = 3;
}

message NodeInfo {
    optional bytes public_ip = 1;
    optional int32 public_port = 2;
    optional bytes local_ip = 3;
    optional int32 local_port = 4;
    optional bytes id = 5;
    optional int32 nat_type = 6;
    optional bytes dht_key = 7;
}

message RefreshNeighborsResponse {
    repeated NodeInfo nodes = 1;
}

message HeartbeatRequest {
	optional uint64 dht_key_hash = 1;
}

message HeartbeatResponse {
	optional uint64 dht_key_hash = 1;
}

message ConnectReqeust {
    optional bytes local_ip = 1;
    optional int32 local_port = 2;
    optional bytes public_ip = 3;
    optional int32 public_port = 4;
    optional int32 nat_type = 5;
	optional bytes id = 6;
    optional bytes dht_key = 7;
	optional bool direct = 8;
}

message DhtMessage {
	optional BootstrapRequest bootstrap_req = 1;
	optional BootstrapResponse bootstrap_res = 2;
	optional RefreshNeighborsRequest refresh_neighbors_req = 3;
	optional RefreshNeighborsResponse refresh_neighbors_res = 4;
	optional ConnectReqeust connect_req = 5;
	optional HeartbeatRequest heartbeat_req = 6;
	optional HeartbeatResponse heartbeat_res = 7;
}